### PR TITLE
Drop most `CardanoApi.Tx` usages in `balanceTx`.

### PR DESCRIPTION
**Pull request by Anviking**
- [x] Drop most CardanoApi.Tx usages in balanceTx — in particular in `updateTx` and `balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment` 

### Issue Number

ADP-2353

### CODE REVIEW

File: lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs [hunk 1]
```diff
@@ -1265,9 +1270,15 @@ spec_updateTx = describe "updateTx" $ do
             -- Could be argued that it should instead return `Right tx`.
             let anyRecentEraTx = recentEraTxFromBytes
                     $ snd $ head signedTxs
-            Write.withInAnyRecentEra anyRecentEraTx $ \tx ->
-                updateTx tx noTxUpdate
-                    `shouldBe` Left (ErrExistingKeyWitnesses 1)
+            Write.withInAnyRecentEra anyRecentEraTx
+                $ \(tx :: Cardano.Tx era) -> do
+                    let res = toCardanoTx @era
```
Comment (Anviking) [inline lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs]: The new need for type-applications is very ugly, but think we can live with it for a while.

Rewriting these tests for ledger types should fix it.

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs [hunk 2]
```diff
@@ -324,6 +333,9 @@ import qualified Cardano.Wallet.Primitive.Types.UTxO as W
 import qualified Cardano.Wallet.Shelley.Compatibility.Ledger as Convert
 import qualified Data.Foldable as F
 import qualified Data.List as L
+import Data.Map
+    ( Map
+    )
```
Comment (jonathanknowles) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: 💥

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs [hunk 3]
```diff
@@ -827,47 +842,48 @@ balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment
...
-        tx' <- left ErrBalanceTxUpdateError $ updateTx tx update
+        tx' <- left ErrBalanceTxUpdateError $ updateTx era tx update
         let balance = txBalance tx'
             minfee' = Cardano.Lovelace $ W.Coin.toInteger minfee
         return (balance, minfee', witCount)
       where
-        getBody (Cardano.Tx body _) = body
+        cardanoApiTxBody :: Cardano.TxBody era
+        cardanoApiTxBody =
+            let (Cardano.Tx body _) = toCardanoTx tx
```
Comment (jonathanknowles) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: Are the parentheses necessary here?

```suggestion
            let Cardano.Tx body _ = toCardanoTx tx
```

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs [hunk 4]
```diff
@@ -1217,38 +1233,36 @@ newtype ErrUpdateSealedTx
...
 
+    extraInputScripts'
+        :: Map (ScriptHash StandardCrypto) (Script (ShelleyLedgerEra era))
+    extraInputScripts' = withConstraints era $
+        Map.fromList $ map (pairWithHash . convert) extraInputScripts
+      where
+        pairWithHash s = (hashScript s, s)
+        convert = flip toLedgerScript era
+
+
```
Comment (jonathanknowles) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]:
```suggestion

```

**Review by jonathanknowles — COMMENTED** [review]
Hi @Anviking 

Just some early comments.

**Review by jonathanknowles — APPROVED** [review]

**Pull request merged**